openshift/dynamic/client.py

```diff
@@ -403,19 +324,227 @@ def __getattr__(self, name):
+        subresources_raw = list(filter(lambda resource: '/' in resource['name'], resources_response))
+        for subresource in subresources_raw:
+            resource, name = subresource['name'].split('/')
+            if not subresources.get(resource):
+                subresources[resource] = {}
+            subresources[resource][name] = subresource
+
+        for resource in resources_raw:
+            # Prevent duplicate keys
+            for key in ('prefix', 'group', 'api_version', 'client', 'preferred'):
+                resource.pop(key, None)
+
+            resources[resource['kind']] = Resource(
+                prefix=prefix,
+                group=group,
+                api_version=version,
+                client=self.client,
+                preferred=preferred,
+                subresources=subresources.get(resource['name']),
+                **resource
+            )
+        return resources
+
+
+class LazyDiscoverer(Discoverer):
     """ A convenient container for storing discovered API resources. Allows
-        easy searching and retrieval of specific resources
+        easy searching and retrieval of specific resources.
+
+        Resources for the cluster are loaded lazily.
     """
 
-    def __init__(self, resources):
-        self.__resources = resources
+    class ResourceGroup(object):
+        def __init__(self, preferred, resources={}):
+            self.preferred = preferred
+            self.resources = resources
+
+    def __init__(self, client):
+        Discoverer.__init__(self,client)
+        self.__resources = self.__setup_resources()
+
+
+    def api_groups(self):
+        return self.__resources['apis'].keys()
+
+    def get(self, **kwargs):
+        """ Same as search, but will throw an error if there are multiple or no
+            results. If there are multiple results and only one is an exact match
+            on api_version, that resource will be returned.
+        """
+        results = self.search(**kwargs)
+        if len(results) > 1 and kwargs.get('api_version'):
+            results = [
+                result for result in results if result.group_version == kwargs['api_version']
+            ]
+        if len(results) == 1:
+            return results[0]
+        elif not results:
+            raise ResourceNotFoundError('No matches found for {}'.format(kwargs))
+        else:
+            raise ResourceNotUniqueError('Multiple matches found for {}: {}'.format(kwargs, results))
+
+    def search(self, **kwargs):
+        return self.__search(self.__build_search(**kwargs), self.__resources, [])
+
+    def __search(self,  parts, resources, reqParams):
+        part = parts[0]
+        if part != '*':
+            resourcePart = resources.get(part)
+            if not resourcePart:
+                return []
+            elif isinstance(resourcePart, self.ResourceGroup):
+                assert len(reqParams) == 2, "prefix and group params should be present, have %s" % reqParams
```

Comment:
Could you raise an Exception (maybe ValueError or something like that) rather than an assertion error?